Core/Graphics/Source/FrameBuffer.cpp

```diff
@@ -53,37 +53,69 @@ namespace Babylon::Graphics
 
     void FrameBuffer::Clear(bgfx::Encoder& encoder, uint16_t flags, uint32_t rgba, float depth, uint8_t stencil)
     {
-        // BGFX requires us to create a new viewID, this will ensure that the view gets cleaned.
+        // BGFX requires us to create a new viewID, this will ensure that the view gets cleared.
         m_viewId = m_context.AcquireNewViewId(encoder);
 
         bgfx::setViewMode(m_viewId.value(), bgfx::ViewMode::Sequential);
         bgfx::setViewClear(m_viewId.value(), flags, rgba, depth, stencil);
         bgfx::setViewFrameBuffer(m_viewId.value(), m_handle);
 
-        // BGFX will consider the viewport when cleaning the screen, but WebGL always cleans the entire screen.
-        // That's why we always set the viewport to {0, 0, 1, 1} when cleaning.
-        bgfx::setViewRect(m_viewId.value(), 0, 0, Width(), Height());
+        // If a scissor is set, we need to set the BGFX view rect to match it before clearing.
+        // We set the view rect instead of the view scissor because BGFX clears after the view rect is set and before
+        // the view scissor is set.
+        // 
+        // If no scissor is set, WebGL cleans the entire screen, so set the view rect to cover the entire screen before
+        // clearing to match WebGL's behavior; otherwise BGFX will only clear the view rect.
+        //
+        // Note that the view rect and view scissor are reset to the desired dimensions before the encoder is submitted.
+        if (m_desiredScissor.Width != 0.0 || m_desiredScissor.Height != 0.0)
+        {
+            const auto flippedScissor = GetFlippedScissor();
+            bgfx::setViewRect(
+                m_viewId.value(),
+                static_cast<uint16_t>(flippedScissor.X),
+                static_cast<uint16_t>(flippedScissor.Y),
+                static_cast<uint16_t>(flippedScissor.Width),
+                static_cast<uint16_t>(flippedScissor.Height));
+
+            m_bgfxViewPort = {
+                m_desiredScissor.X / Width(),
+                m_desiredScissor.Y / Height(),
+                m_desiredScissor.Width / Width(),
+                m_desiredScissor.Height / Height(),
+            };
+        }
+        else
+        {
+            bgfx::setViewRect(m_viewId.value(), 0, 0, Width(), Height());
+            m_bgfxViewPort = {0, 0, 1, 1};
+        }
         encoder.touch(m_viewId.value());
 
-        m_bgfxViewPort = {0, 0, 1, 1};
     }
 
     void FrameBuffer::SetViewPort(bgfx::Encoder& encoder, float x, float y, float width, float height)
     {
         m_desiredViewPort = {x, y, width, height};
-        SetBgfxViewPort(encoder, m_desiredViewPort);
+        SetBgfxViewPortAndScissor(encoder, m_desiredViewPort, m_desiredScissor);
+    }
+
+    void FrameBuffer::SetScissor(bgfx::Encoder& encoder, float x, float y, float width, float height)
+    {
+        m_desiredScissor = {x, y, width, height};
```

Comment:
Is this in a different unit than view port. View port is normalized 0 to 1. Is scissor not normalized? If not normalized, how does hardware scaling affect this?

Reply:
Scissor is not normalized and is not affected by hardware scaling when I change it with `Engine.setHardwareScalingLevel`.

Reply:
It seems the code is using `Width()` and `Height()` to do some calculations relating to scissor rects. `Width()` and `Height()` do take hardware scaling into account. Is that an issue?

Reply:
Maybe we can add a test with non-default hardware scaling with scissor rects also?

Reply:
Maybe I'm confused. What's the test for this? I already tested this by setting `engine.setHardwareScalingLevel` to `0.5` and `0.25`, and the scissor rect took up the same amount of space on the screen for each level. Am I missing something?

Reply:
If that's the case, then hardware scaling is being taken into account.

> Scissor is not normalized and is not affected by hardware scaling when I change it with `Engine.setHardwareScalingLevel`

Maybe I'm not interpreting this statement correctly.

Reply:
Just to clarify, we are sure that this code works correctly if we set hardware scaling on the engine?

Reply:
Ya, the code works correctly with hardware scaling. I checked it manually. There is no visualization test for this, yet.

Reply:
Tests added. Note that bgfx screenshots adjust to the hardware scaling level so I updated validation_native.js to match it. This could lead to large output images when the hardware scaling level is small (a 0.25 scaling level will result in a ~15MB .png). I used a 0.9 scaling level to keep the image size down in this PR but maybe there's a better way to keep the .png files small?